#Exercise 1

In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 9.1 MB 4.9 MB/s 
     |████████████████████████████████| 49 kB 5.2 MB/s 


In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
coef = np.loadtxt('/content/lab4_ex1_coef.txt', delimiter = ',')

In [ ]:
print(coef.shape)
print('Num rows:',coef.shape[0])
print('Num cols:',coef.shape[1])

(6, 7)
Num rows: 6
Num cols: 7


In [ ]:
print(coef)

[[  0.85   3.75  -8.75  -3.45 -22.38  20.     0.  ]
 [ -0.7   -3.9   -1.     0.     1.     0.    14.9 ]
 [  1.     0.     0.    -1.     0.    -1.     0.9 ]
 [  0.    -3.    -2.     0.     0.    -1.   -51.7 ]
 [  0.     0.     0.     4.8    9.12   7.2   75.  ]
 [ -0.8   -1.5    8.    -2.     1.     0.   -27.  ]]


In [ ]:
model = ConcreteModel()

In [ ]:
N = coef.shape[1]-1
M = coef.shape[0]-1

In [ ]:
obj_coef = coef[0][:-1]
print(obj_coef.shape)
print(obj_coef)

(6,)
[  0.85   3.75  -8.75  -3.45 -22.38  20.  ]


In [ ]:
constr_coef = coef[1:,:-1]
print(constr_coef.shape)
print(constr_coef)

(5, 6)
[[-0.7  -3.9  -1.    0.    1.    0.  ]
 [ 1.    0.    0.   -1.    0.   -1.  ]
 [ 0.   -3.   -2.    0.    0.   -1.  ]
 [ 0.    0.    0.    4.8   9.12  7.2 ]
 [-0.8  -1.5   8.   -2.    1.    0.  ]]


In [ ]:
constr_rhs = coef[1:,-1]
print(constr_rhs.shape)
print(constr_rhs)

(5,)
[ 14.9   0.9 -51.7  75.  -27. ]


In [ ]:
lower_bound = 0
upper_bound = np.inf

In [ ]:
col_indices = np.arange(N)
model.x = Var(col_indices)

In [ ]:
row_indices = np.arange(M)
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model.constraints.add(summation(constr_coef[i],model.x)<=constr_rhs[i])

In [ ]:
for j in col_indices:
  model.x[j].setlb(lower_bound)
  model.x[j].setub(upper_bound)

In [ ]:
model.objective = Objective(expr = summation(obj_coef,model.x),sense = maximize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :   inf : False :  True :  Reals
          1 :     0 :  None :   inf : False :  True :  Reals
          2 :     0 :  None :   inf : False :  True :  Reals
          3 :     0 :  None :   inf : False :  True :  Reals
          4 :     0 :  None :   inf : False :  True :  Reals
          5 :     0 :  None :   inf : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : maximize : 0.85*x[0] + 3.75*x[1] - 8.75*x[2

In [ ]:
!apt-get install -y -qq coinor-cbc

Selecting previously unselected package coinor-libcoinutils3v5.
(Reading database ... 148486 files and directories currently installed.)
Preparing to unpack .../0-coinor-libcoinutils3v5_2.10.14+repack1-1_amd64.deb ...
Unpacking coinor-libcoinutils3v5 (2.10.14+repack1-1) ...
Selecting previously unselected package coinor-libosi1v5.
Preparing to unpack .../1-coinor-libosi1v5_0.107.9+repack1-1_amd64.deb ...
Unpacking coinor-libosi1v5 (0.107.9+repack1-1) ...
Selecting previously unselected package coinor-libclp1.
Preparing to unpack .../2-coinor-libclp1_1.16.11+repack1-1_amd64.deb ...
Unpacking coinor-libclp1 (1.16.11+repack1-1) ...
Selecting previously unselected package coinor-libcgl1.
Preparing to unpack .../3-coinor-libcgl1_0.59.10+repack1-1_amd64.deb ...
Unpacking coinor-libcgl1 (0.59.10+repack1-1) ...
Selecting previously unselected package coinor-libcbc3.
Preparing to unpack .../4-coinor-libcbc3_2.9.9+repack1-1_amd64.deb ...
Unpacking coinor-libcbc3 (2.9.9+repack1-1) ...
Selecting p

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

    model.name="unknown";
      - termination condition: unbounded
      - message from solver: <undefined>
Solver status: warning
Solver termination condition: unbounded


## Ans 7 :

The Solver Status gives a warning and the Solver Termination Condition comes out to be unbounded.
This happens because there is no upperbound on the variables and we are maximising the problem.

In [ ]:
model.objective.set_sense(minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    5 : {1, 2, 3, 4, 5}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    6 : {0, 1, 2, 3, 4, 5}

1 Var Declarations
    x : Size=6, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :   inf : False :  True :  Reals
          1 :     0 :  None :   inf : False :  True :  Reals
          2 :     0 :  None :   inf : False :  True :  Reals
          3 :     0 :  None :   inf : False :  True :  Reals
          4 :     0 :  None :   inf : False :  True :  Reals
          5 :     0 :  None :   inf : False :  True :  Reals

1 Objective Declarations
    objective : Size=1, Index=None, Active=True
        Key  : Active : Sense    : Expression
        None :   True : minimize : 0.85*x[0] + 3.75*x[1] - 8.75*x[2

In [ ]:
result = opt_cbc.solve(model)
print('Solver status:', result.solver.status)
print('Solver termination condition:',result.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('Objective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()

Objective =  -97.02184239599997

Decision Variables
x[ 0 ] =  0.9
x[ 1 ] =  23.002456
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  8.2236842
x[ 5 ] =  0.0

Constraints
constraints : Size=5
    Key : Lower : Body                : Upper
      1 :  None :         -82.1158942 :  14.9
      2 :  None :                 0.9 :   0.9
      3 :  None :          -69.007368 : -51.7
      4 :  None :   74.99999990399999 :  75.0
      5 :  None : -26.999999799999998 : -27.0


##Ans 9 :

To get an optimal value we change the sense from maximization to minimization.

The optimal objective function value is : -97.02184239599997

The values of decision variables are as follows:

$x_1 = 0.9\\x_2 = 23.002456\\x_3 = 0.0\\x_4 = 0.0\\x_5 = 8.2236842\\x_6 = 0.0$

The constraint 1 : $-0.7x_1-3.9x_2-x_3+x_5\leq14.9$ is equal to -82.1158942 and hence is not satisfied to equality.

The constraint 2 : $x_1-x_4-x_6\leq0.9$ is equal to 0.9 and hence is satisfied to equality.

The constraint 3 : $-3x_2-2x_3-x_6\leq-51.7$ is equal to -69.007368 and hence is not satisfied to equality.

The constraint 4 : $4.8x_4+9.12x_5+7.2x_6\leq75$ is equal to 74.9999 and hence is not satisfied to equality.

The constraint 5 : $-0.8x_1-1.5x_2+8x_3-2x_4+x_5\leq-27$ is equal to -26.9999 and hence is not satisfied to equality.